{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a55e0a69",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "from helpers.cell_type_naming import nice_to_weirds, weird_to_nice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a4d3ba7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "\n",
    "logging.getLogger('root').addHandler(logging.NullHandler())\n",
    "\n",
    "# configure a handler\n",
    "format_str = '%(asctime)s - %(name)s - %(levelname)s - %(message)s'\n",
    "handler = logging.StreamHandler()\n",
    "handler.setFormatter(logging.Formatter(format_str))\n",
    "\n",
    "# make a logger for this notebook\n",
    "logger = logging.getLogger(__name__)\n",
    "logger.setLevel(\"DEBUG\")\n",
    "logger.addHandler(handler)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6cf45853",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-10-07 17:34:42,210 - __main__ - DEBUG - reading single cells used for sigmat generation\n"
     ]
    }
   ],
   "source": [
    "logger.debug(\"reading single cells used for sigmat generation\")\n",
    "sc_refsample = pd.read_csv(\n",
    "    \"gs://liulab/csx_example_files/Single_Cell_RNA-Seq_Melanoma_SuppFig_3b-d/scRNA-Seq_reference_melanoma_Tirosh_SuppFig_3b-d.txt\",\n",
    "    sep=\"\\t\",\n",
    "    index_col=0,\n",
    "#     nrows=1000  # for debugging\n",
    "#     skiprows=lambda i: i % 500\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c31bf8cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-10-07 17:34:45,693 - __main__ - DEBUG - loading original column names (which have duplicates so pandas messes them up)\n",
      "2021-10-07 17:34:45,943 - __main__ - DEBUG - replacing column names with good ones\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>...</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "      <th>B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C9orf152</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPS11</th>\n",
       "      <td>135.788203</td>\n",
       "      <td>153.354101</td>\n",
       "      <td>296.923889</td>\n",
       "      <td>283.818688</td>\n",
       "      <td>313.256767</td>\n",
       "      <td>323.528192</td>\n",
       "      <td>292.029073</td>\n",
       "      <td>320.594740</td>\n",
       "      <td>148.106568</td>\n",
       "      <td>168.636169</td>\n",
       "      <td>...</td>\n",
       "      <td>222.851783</td>\n",
       "      <td>620.452365</td>\n",
       "      <td>308.307928</td>\n",
       "      <td>969.846904</td>\n",
       "      <td>214.314251</td>\n",
       "      <td>418.579389</td>\n",
       "      <td>263.184131</td>\n",
       "      <td>327.511485</td>\n",
       "      <td>321.778349</td>\n",
       "      <td>840.663649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ELMO2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.149704</td>\n",
       "      <td>0.357997</td>\n",
       "      <td>5.159017</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.927005</td>\n",
       "      <td>4.695806</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.274004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7.250034</td>\n",
       "      <td>9.348143</td>\n",
       "      <td>0.886994</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CREB3L1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PNMA1</th>\n",
       "      <td>1.780029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.870904</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3IP1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.345004</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNRPD2</th>\n",
       "      <td>56.004346</td>\n",
       "      <td>25.364048</td>\n",
       "      <td>82.975699</td>\n",
       "      <td>51.746373</td>\n",
       "      <td>171.888728</td>\n",
       "      <td>91.655197</td>\n",
       "      <td>71.564901</td>\n",
       "      <td>66.761786</td>\n",
       "      <td>82.349405</td>\n",
       "      <td>137.131639</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>69.063930</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>23.430306</td>\n",
       "      <td>45.850742</td>\n",
       "      <td>65.312089</td>\n",
       "      <td>28.601533</td>\n",
       "      <td>5.083915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC39A6</th>\n",
       "      <td>30.710748</td>\n",
       "      <td>17.722485</td>\n",
       "      <td>17.310509</td>\n",
       "      <td>6.795816</td>\n",
       "      <td>12.853705</td>\n",
       "      <td>7.719830</td>\n",
       "      <td>28.671377</td>\n",
       "      <td>2.438073</td>\n",
       "      <td>6.099906</td>\n",
       "      <td>7.051737</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.554878</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTSC</th>\n",
       "      <td>15.550467</td>\n",
       "      <td>4.543127</td>\n",
       "      <td>7.130252</td>\n",
       "      <td>15.113522</td>\n",
       "      <td>5.018901</td>\n",
       "      <td>1.876072</td>\n",
       "      <td>22.271624</td>\n",
       "      <td>21.019350</td>\n",
       "      <td>6.805007</td>\n",
       "      <td>14.943539</td>\n",
       "      <td>...</td>\n",
       "      <td>1.770987</td>\n",
       "      <td>0.859997</td>\n",
       "      <td>0.403997</td>\n",
       "      <td>7.064024</td>\n",
       "      <td>3.321104</td>\n",
       "      <td>0.477000</td>\n",
       "      <td>1.457004</td>\n",
       "      <td>6.552739</td>\n",
       "      <td>1.189043</td>\n",
       "      <td>2.990031</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23684 rows × 737 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Malignant   Malignant   Malignant   Malignant   Malignant  \\\n",
       "Gene                                                                   \n",
       "C9orf152    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "RPS11     135.788203  153.354101  296.923889  283.818688  313.256767   \n",
       "ELMO2       0.000000   13.149704    0.357997    5.159017    0.000000   \n",
       "CREB3L1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "PNMA1       1.780029    0.000000    0.000000    0.000000    0.000000   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "PIK3IP1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "SNRPD2     56.004346   25.364048   82.975699   51.746373  171.888728   \n",
       "SLC39A6    30.710748   17.722485   17.310509    6.795816   12.853705   \n",
       "CTSC       15.550467    4.543127    7.130252   15.113522    5.018901   \n",
       "AQP7        0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "           Malignant   Malignant   Malignant   Malignant   Malignant  ...  \\\n",
       "Gene                                                                  ...   \n",
       "C9orf152    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "RPS11     323.528192  292.029073  320.594740  148.106568  168.636169  ...   \n",
       "ELMO2       0.927005    4.695806    0.000000    0.000000    0.000000  ...   \n",
       "CREB3L1     0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "PNMA1       0.000000    0.000000    0.000000    9.870904    0.000000  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "PIK3IP1     0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "SNRPD2     91.655197   71.564901   66.761786   82.349405  137.131639  ...   \n",
       "SLC39A6     7.719830   28.671377    2.438073    6.099906    7.051737  ...   \n",
       "CTSC        1.876072   22.271624   21.019350    6.805007   14.943539  ...   \n",
       "AQP7        0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "\n",
       "                   B           B           B           B           B  \\\n",
       "Gene                                                                   \n",
       "C9orf152    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "RPS11     222.851783  620.452365  308.307928  969.846904  214.314251   \n",
       "ELMO2       0.000000    0.000000    0.000000    0.274004    0.000000   \n",
       "CREB3L1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "PNMA1       0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "PIK3IP1     0.000000    0.345004    0.000000    0.000000    0.000000   \n",
       "SNRPD2      0.000000    0.000000   69.063930    0.000000    0.000000   \n",
       "SLC39A6     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "CTSC        1.770987    0.859997    0.403997    7.064024    3.321104   \n",
       "AQP7        0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "                   B           B           B           B           B  \n",
       "Gene                                                                  \n",
       "C9orf152    0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "RPS11     418.579389  263.184131  327.511485  321.778349  840.663649  \n",
       "ELMO2       7.250034    9.348143    0.886994    0.000000    0.000000  \n",
       "CREB3L1     0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "PNMA1       0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "...              ...         ...         ...         ...         ...  \n",
       "PIK3IP1     0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "SNRPD2     23.430306   45.850742   65.312089   28.601533    5.083915  \n",
       "SLC39A6     0.000000    2.554878    0.000000    0.000000    0.000000  \n",
       "CTSC        0.477000    1.457004    6.552739    1.189043    2.990031  \n",
       "AQP7        0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "\n",
       "[23684 rows x 737 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "logger.debug(\"loading original column names (which have duplicates so pandas messes them up)\")\n",
    "from google.cloud import storage\n",
    "client = storage.Client()\n",
    "bucket = client.bucket(\"liulab\")\n",
    "blob = bucket.get_blob(\"csx_example_files/Single_Cell_RNA-Seq_Melanoma_SuppFig_3b-d/scRNA-Seq_reference_melanoma_Tirosh_SuppFig_3b-d.txt\")\n",
    "with blob.open(\"rt\") as f:\n",
    "    original_column_names = f.readline().strip().split(\"\\t\")\n",
    "\n",
    "logger.debug(\"replacing column names with good ones\")\n",
    "good_column_names = [weird_to_nice.get(column, column) for column in original_column_names]\n",
    "sc_refsample.columns = good_column_names[1:]  # exclude index column name\n",
    "\n",
    "sc_refsample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "03a43489",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Malignant', 'Endothelial', 'CAF', 'T CD8', 'NK', 'Macrophage', 'T CD4', 'B']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_types = list(nice_to_weirds)\n",
    "cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5391f8fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-r--r-- 1 jupyter jupyter 28141097 Jun 25 20:59 /mnt/buckets/liulab/derek/simulations/experiments/cibersortx_sim_0.5sd.txt\n",
      "-rw-r--r-- 1 jupyter jupyter 14111 Jun 25 20:58 /mnt/buckets/liulab/derek/simulations/experiments/generate_cohorts_for_emma/ctp_sim_0.5sd.txt\n",
      "-rw-r--r-- 1 jupyter jupyter 28125676 Jun 25 20:58 /mnt/buckets/liulab/derek/simulations/experiments/generate_cohorts_for_emma/sim_0.5sd.txt\n"
     ]
    }
   ],
   "source": [
    "!find /mnt/buckets/liulab/derek/ -type f -name \"*0.5sd*\" -exec ls -l {} \\;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "96947230",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/cibersortx_sim_0.5sd.txt#1633627841575500...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/sc_refsample_modified.txt#1633627840517129...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_CAF/cibersortx_sim_0.5sd.txt#1633627791955481...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_CAF/sc_refsample_modified.txt#1633627790870552...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Endothelial/cibersortx_sim_0.5sd.txt#1633627781360649...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Endothelial/sc_refsample_modified.txt#1633627780298328...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Macrophage/cibersortx_sim_0.5sd.txt#1633627823041665...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Macrophage/sc_refsample_modified.txt#1633627821955022...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Malignant/cibersortx_sim_0.5sd.txt#1633627770758733...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Malignant/sc_refsample_modified.txt#1633627769652343...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_NK/cibersortx_sim_0.5sd.txt#1633627812522392...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_NK/sc_refsample_modified.txt#1633627811442449...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD4/cibersortx_sim_0.5sd.txt#1633627831368262...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD4/sc_refsample_modified.txt#1633627830279329...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD8/cibersortx_sim_0.5sd.txt#1633627801427995...\n",
      "Removing gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD8/sc_refsample_modified.txt#1633627800361230...\n",
      "/ [16/16 objects] 100% Done                                                     \n",
      "Operation completed over 16 objects.                                             \n"
     ]
    }
   ],
   "source": [
    "!gsutil -m rm -r gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "047530f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-10-07 17:34:48,584 - __main__ - DEBUG - exluding Malignant\n",
      "2021-10-07 17:34:48,585 - __main__ - DEBUG - limiting to 495 single cells\n",
      "2021-10-07 17:34:48,585 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Malignant/sc_refsample_modified.txt\n",
      "2021-10-07 17:34:55,868 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Malignant/cibersortx_sim_0.5sd.txt\n",
      "2021-10-07 17:34:55,877 - __main__ - DEBUG - exluding Endothelial\n",
      "2021-10-07 17:34:55,878 - __main__ - DEBUG - limiting to 708 single cells\n",
      "2021-10-07 17:34:55,879 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Endothelial/sc_refsample_modified.txt\n",
      "2021-10-07 17:35:06,490 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Endothelial/cibersortx_sim_0.5sd.txt\n",
      "2021-10-07 17:35:06,500 - __main__ - DEBUG - exluding CAF\n",
      "2021-10-07 17:35:06,501 - __main__ - DEBUG - limiting to 712 single cells\n",
      "2021-10-07 17:35:06,502 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_CAF/sc_refsample_modified.txt\n",
      "2021-10-07 17:35:17,366 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_CAF/cibersortx_sim_0.5sd.txt\n",
      "2021-10-07 17:35:17,376 - __main__ - DEBUG - exluding T CD8\n",
      "2021-10-07 17:35:17,377 - __main__ - DEBUG - limiting to 623 single cells\n",
      "2021-10-07 17:35:17,378 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD8/sc_refsample_modified.txt\n",
      "2021-10-07 17:35:27,025 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD8/cibersortx_sim_0.5sd.txt\n",
      "2021-10-07 17:35:27,034 - __main__ - DEBUG - exluding NK\n",
      "2021-10-07 17:35:27,036 - __main__ - DEBUG - limiting to 724 single cells\n",
      "2021-10-07 17:35:27,036 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_NK/sc_refsample_modified.txt\n",
      "2021-10-07 17:35:38,223 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_NK/cibersortx_sim_0.5sd.txt\n",
      "2021-10-07 17:35:38,232 - __main__ - DEBUG - exluding Macrophage\n",
      "2021-10-07 17:35:38,233 - __main__ - DEBUG - limiting to 697 single cells\n",
      "2021-10-07 17:35:38,234 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Macrophage/sc_refsample_modified.txt\n",
      "2021-10-07 17:35:48,799 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Macrophage/cibersortx_sim_0.5sd.txt\n",
      "2021-10-07 17:35:48,808 - __main__ - DEBUG - exluding T CD4\n",
      "2021-10-07 17:35:48,809 - __main__ - DEBUG - limiting to 528 single cells\n",
      "2021-10-07 17:35:48,810 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD4/sc_refsample_modified.txt\n",
      "2021-10-07 17:35:56,980 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD4/cibersortx_sim_0.5sd.txt\n",
      "2021-10-07 17:35:56,989 - __main__ - DEBUG - exluding B\n",
      "2021-10-07 17:35:56,990 - __main__ - DEBUG - limiting to 672 single cells\n",
      "2021-10-07 17:35:56,991 - __main__ - DEBUG - saving to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/sc_refsample_modified.txt\n",
      "2021-10-07 17:36:07,233 - __main__ - DEBUG - copying mixture file to gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/cibersortx_sim_0.5sd.txt\n"
     ]
    }
   ],
   "source": [
    "gsutil_calls = list()\n",
    "\n",
    "for cell_type in cell_types:\n",
    "    logger.debug(f\"exluding {cell_type}\")\n",
    "    desired_column_positions = [\n",
    "        i for (i, column_name) in enumerate(sc_refsample.columns)\n",
    "        if cell_type not in column_name]\n",
    "    logger.debug(f\"limiting to {len(desired_column_positions)} single cells\")\n",
    "    base_path = f\"gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_{cell_type}\".replace(\" \", \"_\")\n",
    "    filepath_sc_refsample = os.path.join(base_path, \"sc_refsample_modified.txt\")\n",
    "    logger.debug(f\"saving to {filepath_sc_refsample}\")\n",
    "    sc_refsample.iloc[:, desired_column_positions].to_csv(filepath_sc_refsample, sep=\"\\t\")\n",
    "    filepath_mixture = os.path.join(base_path, \"cibersortx_sim_0.5sd.txt\")\n",
    "    logger.debug(f\"copying mixture file to {filepath_mixture}\")\n",
    "    copy_mixture_command = f\"gsutil cp gs://liulab/derek/simulations/experiments/cibersortx_sim_0.5sd.txt {filepath_mixture}\"\n",
    "    gsutil_calls.append(subprocess.Popen(copy_mixture_command, shell=True))\n",
    "\n",
    "for call in gsutil_calls:\n",
    "    call.wait()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "27f09ff9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>Malignant</th>\n",
       "      <th>...</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "      <th>T CD4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C9orf152</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPS11</th>\n",
       "      <td>135.788203</td>\n",
       "      <td>153.354101</td>\n",
       "      <td>296.923889</td>\n",
       "      <td>283.818688</td>\n",
       "      <td>313.256767</td>\n",
       "      <td>323.528192</td>\n",
       "      <td>292.029073</td>\n",
       "      <td>320.594740</td>\n",
       "      <td>148.106568</td>\n",
       "      <td>168.636169</td>\n",
       "      <td>...</td>\n",
       "      <td>303.669232</td>\n",
       "      <td>560.211204</td>\n",
       "      <td>385.868646</td>\n",
       "      <td>439.500547</td>\n",
       "      <td>341.248403</td>\n",
       "      <td>563.879761</td>\n",
       "      <td>467.790655</td>\n",
       "      <td>1045.241365</td>\n",
       "      <td>372.017628</td>\n",
       "      <td>64.132243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ELMO2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.149704</td>\n",
       "      <td>0.357997</td>\n",
       "      <td>5.159017</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.927005</td>\n",
       "      <td>4.695806</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>20.034710</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>128.490280</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CREB3L1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.101035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PNMA1</th>\n",
       "      <td>1.780029</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9.870904</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>22.544203</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>59.171149</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3IP1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37.282151</td>\n",
       "      <td>43.988988</td>\n",
       "      <td>13.487131</td>\n",
       "      <td>44.727812</td>\n",
       "      <td>2.069918</td>\n",
       "      <td>66.130664</td>\n",
       "      <td>137.198678</td>\n",
       "      <td>2.439980</td>\n",
       "      <td>18.475656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNRPD2</th>\n",
       "      <td>56.004346</td>\n",
       "      <td>25.364048</td>\n",
       "      <td>82.975699</td>\n",
       "      <td>51.746373</td>\n",
       "      <td>171.888728</td>\n",
       "      <td>91.655197</td>\n",
       "      <td>71.564901</td>\n",
       "      <td>66.761786</td>\n",
       "      <td>82.349405</td>\n",
       "      <td>137.131639</td>\n",
       "      <td>...</td>\n",
       "      <td>79.020625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>71.902683</td>\n",
       "      <td>43.582327</td>\n",
       "      <td>53.712554</td>\n",
       "      <td>51.269592</td>\n",
       "      <td>139.507176</td>\n",
       "      <td>27.936128</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC39A6</th>\n",
       "      <td>30.710748</td>\n",
       "      <td>17.722485</td>\n",
       "      <td>17.310509</td>\n",
       "      <td>6.795816</td>\n",
       "      <td>12.853705</td>\n",
       "      <td>7.719830</td>\n",
       "      <td>28.671377</td>\n",
       "      <td>2.438073</td>\n",
       "      <td>6.099906</td>\n",
       "      <td>7.051737</td>\n",
       "      <td>...</td>\n",
       "      <td>3.013052</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>31.979554</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.784130</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTSC</th>\n",
       "      <td>15.550467</td>\n",
       "      <td>4.543127</td>\n",
       "      <td>7.130252</td>\n",
       "      <td>15.113522</td>\n",
       "      <td>5.018901</td>\n",
       "      <td>1.876072</td>\n",
       "      <td>22.271624</td>\n",
       "      <td>21.019350</td>\n",
       "      <td>6.805007</td>\n",
       "      <td>14.943539</td>\n",
       "      <td>...</td>\n",
       "      <td>0.913999</td>\n",
       "      <td>65.690067</td>\n",
       "      <td>1.826075</td>\n",
       "      <td>9.488235</td>\n",
       "      <td>0.990996</td>\n",
       "      <td>1.069956</td>\n",
       "      <td>9.913939</td>\n",
       "      <td>2.766958</td>\n",
       "      <td>10.463693</td>\n",
       "      <td>84.936553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AQP7</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.205033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23684 rows × 672 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Malignant   Malignant   Malignant   Malignant   Malignant  \\\n",
       "Gene                                                                   \n",
       "C9orf152    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "RPS11     135.788203  153.354101  296.923889  283.818688  313.256767   \n",
       "ELMO2       0.000000   13.149704    0.357997    5.159017    0.000000   \n",
       "CREB3L1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "PNMA1       1.780029    0.000000    0.000000    0.000000    0.000000   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "PIK3IP1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "SNRPD2     56.004346   25.364048   82.975699   51.746373  171.888728   \n",
       "SLC39A6    30.710748   17.722485   17.310509    6.795816   12.853705   \n",
       "CTSC       15.550467    4.543127    7.130252   15.113522    5.018901   \n",
       "AQP7        0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "           Malignant   Malignant   Malignant   Malignant   Malignant  ...  \\\n",
       "Gene                                                                  ...   \n",
       "C9orf152    0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "RPS11     323.528192  292.029073  320.594740  148.106568  168.636169  ...   \n",
       "ELMO2       0.927005    4.695806    0.000000    0.000000    0.000000  ...   \n",
       "CREB3L1     0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "PNMA1       0.000000    0.000000    0.000000    9.870904    0.000000  ...   \n",
       "...              ...         ...         ...         ...         ...  ...   \n",
       "PIK3IP1     0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "SNRPD2     91.655197   71.564901   66.761786   82.349405  137.131639  ...   \n",
       "SLC39A6     7.719830   28.671377    2.438073    6.099906    7.051737  ...   \n",
       "CTSC        1.876072   22.271624   21.019350    6.805007   14.943539  ...   \n",
       "AQP7        0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "\n",
       "               T CD4       T CD4       T CD4       T CD4       T CD4  \\\n",
       "Gene                                                                   \n",
       "C9orf152    0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "RPS11     303.669232  560.211204  385.868646  439.500547  341.248403   \n",
       "ELMO2       0.000000    0.000000    0.000000   20.034710    0.000000   \n",
       "CREB3L1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "PNMA1       0.000000    0.000000    0.000000   22.544203    0.000000   \n",
       "...              ...         ...         ...         ...         ...   \n",
       "PIK3IP1     0.000000   37.282151   43.988988   13.487131   44.727812   \n",
       "SNRPD2     79.020625    0.000000    0.000000   71.902683   43.582327   \n",
       "SLC39A6     3.013052    0.000000    0.000000   31.979554    0.000000   \n",
       "CTSC        0.913999   65.690067    1.826075    9.488235    0.990996   \n",
       "AQP7        0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "               T CD4       T CD4        T CD4       T CD4      T CD4  \n",
       "Gene                                                                  \n",
       "C9orf152    0.000000    0.000000     0.000000    0.000000   0.000000  \n",
       "RPS11     563.879761  467.790655  1045.241365  372.017628  64.132243  \n",
       "ELMO2       0.000000    0.000000   128.490280    0.000000   0.000000  \n",
       "CREB3L1     0.000000    0.000000     0.000000    0.000000   1.101035  \n",
       "PNMA1      59.171149    0.000000     0.000000    0.000000   0.000000  \n",
       "...              ...         ...          ...         ...        ...  \n",
       "PIK3IP1     2.069918   66.130664   137.198678    2.439980  18.475656  \n",
       "SNRPD2     53.712554   51.269592   139.507176   27.936128   0.000000  \n",
       "SLC39A6     0.000000    4.784130     0.000000    0.000000   0.000000  \n",
       "CTSC        1.069956    9.913939     2.766958   10.463693  84.936553  \n",
       "AQP7        0.000000    0.000000     0.000000    0.000000   1.205033  \n",
       "\n",
       "[23684 rows x 672 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc_refsample.iloc[:, desired_column_positions]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "15cea51a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m/mnt/buckets/liulab/csx_experiments_excluding_cell_types_w_absolute_counts/\u001b[00m\n",
      "├── [   0]  \u001b[01;34mwithout_B\u001b[00m\n",
      "│   ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "│   └── [105M]  sc_refsample_modified.txt\n",
      "├── [   0]  \u001b[01;34mwithout_CAF\u001b[00m\n",
      "│   ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "│   └── [110M]  sc_refsample_modified.txt\n",
      "├── [   0]  \u001b[01;34mwithout_Endothelial\u001b[00m\n",
      "│   ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "│   └── [110M]  sc_refsample_modified.txt\n",
      "├── [   0]  \u001b[01;34mwithout_Macrophage\u001b[00m\n",
      "│   ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "│   └── [108M]  sc_refsample_modified.txt\n",
      "├── [   0]  \u001b[01;34mwithout_Malignant\u001b[00m\n",
      "│   ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "│   └── [ 72M]  sc_refsample_modified.txt\n",
      "├── [   0]  \u001b[01;34mwithout_NK\u001b[00m\n",
      "│   ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "│   └── [112M]  sc_refsample_modified.txt\n",
      "├── [   0]  \u001b[01;34mwithout_T_CD4\u001b[00m\n",
      "│   ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "│   └── [ 84M]  sc_refsample_modified.txt\n",
      "└── [   0]  \u001b[01;34mwithout_T_CD8\u001b[00m\n",
      "    ├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "    └── [ 97M]  sc_refsample_modified.txt\n",
      "\n",
      "8 directories, 16 files\n"
     ]
    }
   ],
   "source": [
    "!tree -h /mnt/buckets/liulab/csx_experiments_excluding_cell_types_w_absolute_counts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3738dbb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/:\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/:\n",
      " 26.84 MiB  2021-10-07T17:36:08Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/cibersortx_sim_0.5sd.txt\n",
      "104.68 MiB  2021-10-07T17:36:07Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/sc_refsample_modified.txt\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_CAF/:\n",
      " 26.84 MiB  2021-10-07T17:35:18Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_CAF/cibersortx_sim_0.5sd.txt\n",
      "109.52 MiB  2021-10-07T17:35:17Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_CAF/sc_refsample_modified.txt\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Endothelial/:\n",
      " 26.84 MiB  2021-10-07T17:35:07Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Endothelial/cibersortx_sim_0.5sd.txt\n",
      "109.68 MiB  2021-10-07T17:35:06Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Endothelial/sc_refsample_modified.txt\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Macrophage/:\n",
      " 26.84 MiB  2021-10-07T17:35:49Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Macrophage/cibersortx_sim_0.5sd.txt\n",
      "107.53 MiB  2021-10-07T17:35:48Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Macrophage/sc_refsample_modified.txt\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Malignant/:\n",
      " 26.84 MiB  2021-10-07T17:34:56Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Malignant/cibersortx_sim_0.5sd.txt\n",
      " 72.23 MiB  2021-10-07T17:34:55Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_Malignant/sc_refsample_modified.txt\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_NK/:\n",
      " 26.84 MiB  2021-10-07T17:35:39Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_NK/cibersortx_sim_0.5sd.txt\n",
      "112.11 MiB  2021-10-07T17:35:38Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_NK/sc_refsample_modified.txt\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD4/:\n",
      " 26.84 MiB  2021-10-07T17:35:58Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD4/cibersortx_sim_0.5sd.txt\n",
      " 84.46 MiB  2021-10-07T17:35:56Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD4/sc_refsample_modified.txt\n",
      "\n",
      "gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD8/:\n",
      " 26.84 MiB  2021-10-07T17:35:28Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD8/cibersortx_sim_0.5sd.txt\n",
      " 97.45 MiB  2021-10-07T17:35:26Z  gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_T_CD8/sc_refsample_modified.txt\n",
      "TOTAL: 16 objects, 1061531612 bytes (1012.36 MiB)\n"
     ]
    }
   ],
   "source": [
    "!gsutil ls -hlR gs://liulab/csx_experiments_excluding_cell_types_w_absolute_counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4814d92",
   "metadata": {},
   "source": [
    "# run CIBERSORTx on these folders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "718b814b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2021-10-07 17:36:10,328 - __main__ - DEBUG - setting up docker command for Malignant\n",
      "2021-10-07 17:36:10,550 - __main__ - DEBUG - calling docker run for Malignant\n",
      "2021-10-07 17:36:11,170 - __main__ - DEBUG - setting up docker command for Endothelial\n",
      "2021-10-07 17:36:11,366 - __main__ - DEBUG - calling docker run for Endothelial\n",
      "2021-10-07 17:36:11,942 - __main__ - DEBUG - setting up docker command for CAF\n",
      "2021-10-07 17:36:12,126 - __main__ - DEBUG - calling docker run for CAF\n",
      "2021-10-07 17:36:12,731 - __main__ - DEBUG - setting up docker command for T CD8\n",
      "2021-10-07 17:36:12,943 - __main__ - DEBUG - calling docker run for T CD8\n",
      "2021-10-07 17:36:13,522 - __main__ - DEBUG - setting up docker command for NK\n",
      "2021-10-07 17:36:13,717 - __main__ - DEBUG - calling docker run for NK\n",
      "2021-10-07 17:36:14,334 - __main__ - DEBUG - setting up docker command for Macrophage\n",
      "2021-10-07 17:36:14,525 - __main__ - DEBUG - calling docker run for Macrophage\n",
      "2021-10-07 17:36:15,134 - __main__ - DEBUG - setting up docker command for T CD4\n",
      "2021-10-07 17:36:15,325 - __main__ - DEBUG - calling docker run for T CD4\n",
      "2021-10-07 17:36:16,119 - __main__ - DEBUG - setting up docker command for B\n",
      "2021-10-07 17:36:16,310 - __main__ - DEBUG - calling docker run for B\n"
     ]
    }
   ],
   "source": [
    "for cell_type in cell_types:\n",
    "    logger.debug(f\"setting up docker command for {cell_type}\")\n",
    "    experiment_path = f\"/mnt/buckets/liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_{cell_type}\".replace(\" \", \"_\")\n",
    "    os.makedirs(os.path.join(experiment_path, \"out\"), exist_ok=True)\n",
    "    command = f'''docker run \\\n",
    "    --rm \\\n",
    "    -d \\\n",
    "    -v {experiment_path}:/src/data \\\n",
    "    -v {experiment_path}/out:/src/outdir \\\n",
    "    --user \"$(id -u):$(id -g)\" \\\n",
    "    cibersortx/fractions:latest \\\n",
    "    --username lyronctk@stanford.edu \\\n",
    "    --token dfeba2c8b9d61daebee5fa87026b8e56 \\\n",
    "    --single_cell TRUE \\\n",
    "    --absolute TRUE \\\n",
    "    --refsample sc_refsample_modified.txt \\\n",
    "    --mixture cibersortx_sim_0.5sd.txt \\\n",
    "    --replicates 5 \\\n",
    "    --sampling 0.5 \\\n",
    "    --fraction 0.75 \\\n",
    "    --k.max 999 \\\n",
    "    --q.value 0.01 \\\n",
    "    --G.min 300 \\\n",
    "    --G.max 500 \\\n",
    "    --filter FALSE \\\n",
    "    --verbose TRUE \\\n",
    "    --QN FALSE\n",
    "    '''\n",
    "    logger.debug(f\"calling docker run for {cell_type}\")\n",
    "    subprocess.run(command, shell=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "08e9486a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CONTAINER ID   IMAGE                          COMMAND                  CREATED         STATUS                  PORTS     NAMES\n",
      "0d40392d5bc6   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   1 second ago    Up Less than a second             bold_bouman\n",
      "c1eb430b0e51   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   2 seconds ago   Up 1 second                       ecstatic_booth\n",
      "a7b465066790   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   3 seconds ago   Up 2 seconds                      adoring_ardinghelli\n",
      "3b5037769591   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   4 seconds ago   Up 2 seconds                      xenodochial_austin\n",
      "ccb920b9f239   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 seconds ago   Up 3 seconds                      inspiring_yalow\n",
      "94df422c952e   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 seconds ago   Up 4 seconds                      angry_banach\n",
      "c27e0f8eeed9   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   6 seconds ago   Up 5 seconds                      nostalgic_sutherland\n",
      "b6fbf83a8a37   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   7 seconds ago   Up 6 seconds                      zen_maxwell\n",
      "14346d93c5eb   gcr.io/inverting-proxy/agent   \"/bin/sh -c '/opt/bi…\"   5 hours ago     Up 5 hours                        proxy-agent\n"
     ]
    }
   ],
   "source": [
    "!docker ps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1380576",
   "metadata": {},
   "source": [
    "# appendix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3ec1b09e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cells</th>\n",
       "      <th>samples</th>\n",
       "      <th>cell.types</th>\n",
       "      <th>treatment.group</th>\n",
       "      <th>Cohort</th>\n",
       "      <th>no.of.genes</th>\n",
       "      <th>no.of.reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cy78_CD45_neg_1_B04_S496_comb</td>\n",
       "      <td>Mel78</td>\n",
       "      <td>Mal</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>8258</td>\n",
       "      <td>357919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb</td>\n",
       "      <td>Mel79</td>\n",
       "      <td>Mal</td>\n",
       "      <td>treatment.naive</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>2047</td>\n",
       "      <td>5727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CY88_5_B10_S694_comb</td>\n",
       "      <td>Mel88</td>\n",
       "      <td>Mal</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>5375</td>\n",
       "      <td>139218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb</td>\n",
       "      <td>Mel79</td>\n",
       "      <td>Mal</td>\n",
       "      <td>treatment.naive</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>5648</td>\n",
       "      <td>73996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cy78_CD45_neg_3_H06_S762_comb</td>\n",
       "      <td>Mel78</td>\n",
       "      <td>Mal</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>7409</td>\n",
       "      <td>380341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7181</th>\n",
       "      <td>CY75_1_CD45_CD8_3__S168_comb_BCD8</td>\n",
       "      <td>Mel75</td>\n",
       "      <td>T.CD8</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>3530</td>\n",
       "      <td>137245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7182</th>\n",
       "      <td>CY75_1_CD45_CD8_8__S338_comb_BCD8</td>\n",
       "      <td>Mel75</td>\n",
       "      <td>T.CD8</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>3872</td>\n",
       "      <td>106432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7183</th>\n",
       "      <td>monika_D7_S132_comb_BCD8_3</td>\n",
       "      <td>Mel75</td>\n",
       "      <td>T.CD8</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>4589</td>\n",
       "      <td>908173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7184</th>\n",
       "      <td>CY75_1_CD45_CD8_8__S289_comb_BCD8</td>\n",
       "      <td>Mel75</td>\n",
       "      <td>T.CD8</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>4614</td>\n",
       "      <td>140903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7185</th>\n",
       "      <td>CY75_1_CD45_CD8_8__S351_comb_BCD8</td>\n",
       "      <td>Mel75</td>\n",
       "      <td>T.CD8</td>\n",
       "      <td>post.treatment</td>\n",
       "      <td>Tirosh</td>\n",
       "      <td>3859</td>\n",
       "      <td>222239</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7186 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                cells samples cell.types  \\\n",
       "0                       cy78_CD45_neg_1_B04_S496_comb   Mel78        Mal   \n",
       "1            cy79_p4_CD45_neg_PDL1_neg_E11_S1115_comb   Mel79        Mal   \n",
       "2                                CY88_5_B10_S694_comb   Mel88        Mal   \n",
       "3     cy79_p1_CD45_neg_PDL1_pos_AS_C1_R1_F07_S67_comb   Mel79        Mal   \n",
       "4                       cy78_CD45_neg_3_H06_S762_comb   Mel78        Mal   \n",
       "...                                               ...     ...        ...   \n",
       "7181                CY75_1_CD45_CD8_3__S168_comb_BCD8   Mel75      T.CD8   \n",
       "7182                CY75_1_CD45_CD8_8__S338_comb_BCD8   Mel75      T.CD8   \n",
       "7183                       monika_D7_S132_comb_BCD8_3   Mel75      T.CD8   \n",
       "7184                CY75_1_CD45_CD8_8__S289_comb_BCD8   Mel75      T.CD8   \n",
       "7185                CY75_1_CD45_CD8_8__S351_comb_BCD8   Mel75      T.CD8   \n",
       "\n",
       "      treatment.group  Cohort  no.of.genes  no.of.reads  \n",
       "0      post.treatment  Tirosh         8258       357919  \n",
       "1     treatment.naive  Tirosh         2047         5727  \n",
       "2      post.treatment  Tirosh         5375       139218  \n",
       "3     treatment.naive  Tirosh         5648        73996  \n",
       "4      post.treatment  Tirosh         7409       380341  \n",
       "...               ...     ...          ...          ...  \n",
       "7181   post.treatment  Tirosh         3530       137245  \n",
       "7182   post.treatment  Tirosh         3872       106432  \n",
       "7183   post.treatment  Tirosh         4589       908173  \n",
       "7184   post.treatment  Tirosh         4614       140903  \n",
       "7185   post.treatment  Tirosh         3859       222239  \n",
       "\n",
       "[7186 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc_annotations = pd.read_csv(\"gs://liulab/ftp/GSE115978/GSE115978_cell.annotations.csv\")\n",
    "sc_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e3ad08df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cell_types = list(sc_annotations[\"cell.types\"].unique())\n",
    "# cell_types.remove(\"?\")\n",
    "# cell_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dcc02023",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from google.cloud import storage\n",
    "# client = storage.Client()\n",
    "# bucket = client.bucket(\"liulab\")\n",
    "# blob = bucket.get_blob(\"csx_example_files/Single_Cell_RNA-Seq_Melanoma_SuppFig_3b-d/scRNA-Seq_reference_melanoma_Tirosh_SuppFig_3b-d.txt\")\n",
    "# with blob.open(\"rt\") as f:\n",
    "#     original_column_names = f.readline().strip().split(\"\\t\")\n",
    "# melanoma_sigmatrix_cells.columns = original_column_names[1:]  # exclude index column name"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4ab6bb0",
   "metadata": {},
   "source": [
    "# Check one of the results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1875f482",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m/mnt/buckets/liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/\u001b[00m\n",
      "├── [ 27M]  cibersortx_sim_0.5sd.txt\n",
      "├── [   0]  \u001b[01;34mout\u001b[00m\n",
      "│   ├── [3.4M]  CIBERSORTx_cell_type_sourceGEP.txt\n",
      "│   ├── [133K]  CIBERSORTx_sc_refsample_modified_inferred_phenoclasses.CIBERSORTx_sc_refsample_modified_inferred_refsample.bm.K999.pdf\n",
      "│   ├── [381K]  CIBERSORTx_sc_refsample_modified_inferred_phenoclasses.CIBERSORTx_sc_refsample_modified_inferred_refsample.bm.K999.txt\n",
      "│   ├── [ 542]  CIBERSORTx_sc_refsample_modified_inferred_phenoclasses.txt\n",
      "│   ├── [ 16M]  CIBERSORTx_sc_refsample_modified_inferred_refsample.txt\n",
      "│   ├── [365K]  temp.Fractions.coreSVR.X.tsv\n",
      "│   ├── [5.1M]  temp.Fractions.coreSVR.Y.tsv\n",
      "│   └── [  59]  temp.Fractions.coreSVR.nus.tsv\n",
      "└── [105M]  sc_refsample_modified.txt\n",
      "\n",
      "1 directory, 10 files\n"
     ]
    }
   ],
   "source": [
    "!tree -h /mnt/buckets/liulab/csx_experiments_excluding_cell_types_w_absolute_counts/without_B/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0c83fe17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CONTAINER ID   IMAGE                          COMMAND                  CREATED         STATUS         PORTS     NAMES\n",
      "0d40392d5bc6   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             bold_bouman\n",
      "c1eb430b0e51   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             ecstatic_booth\n",
      "a7b465066790   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             adoring_ardinghelli\n",
      "3b5037769591   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             xenodochial_austin\n",
      "ccb920b9f239   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             inspiring_yalow\n",
      "94df422c952e   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             angry_banach\n",
      "c27e0f8eeed9   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             nostalgic_sutherland\n",
      "b6fbf83a8a37   cibersortx/fractions:latest    \"./CIBERSORTxFractio…\"   5 minutes ago   Up 5 minutes             zen_maxwell\n",
      "14346d93c5eb   gcr.io/inverting-proxy/agent   \"/bin/sh -c '/opt/bi…\"   5 hours ago     Up 5 hours               proxy-agent\n"
     ]
    }
   ],
   "source": [
    "!docker ps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d2582d6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "tf2-gpu.2-5.m76",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/tf2-gpu.2-5:m76"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
