{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# analyze false positive genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "what explains false positives?\n",
    "\n",
    "for each gene:\n",
    "- get p-values, fold changes predicted by cibersortx\n",
    "  - load from pre-computed results\n",
    "- compute ratio of mean expression in malignant cells vs immune cells\n",
    "\n",
    "make plots\n",
    "- plot scatter: \n",
    "  - expression ratio vs average FPR (across trials)\n",
    "  - expression ratio vs average p-value\n",
    "\n",
    "looking for:\n",
    "- does low malignant ratio predict false discovery?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import helpers\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import upath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[GCSPath('gs://liulab/differential_composition/20230120_04h22m54s/run_id=00/malignant_means=0.55,0.85/deg_analysis/gene_stats_malignant_cibersortx.parquet'), GCSPath('gs://liulab/differential_composition/20230120_04h22m54s/run_id=00/malignant_means=0.6,0.8/deg_analysis/gene_stats_malignant_cibersortx.parquet'), GCSPath('gs://liulab/differential_composition/20230120_04h22m54s/run_id=00/malignant_means=0.65,0.75/deg_analysis/gene_stats_malignant_cibersortx.parquet')]\n"
     ]
    }
   ],
   "source": [
    "path_root = upath.UPath(\"gs://liulab/differential_composition/20230120_04h22m54s\")\n",
    "parquet_paths = list(\n",
    "    path_root.glob(\"*/malignant_means=*/deg_analysis/gene_stats_malignant_cibersortx.parquet\")\n",
    ")\n",
    "print(parquet_paths[:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_from_path(path: str, var_name: str) -> str:\n",
    "    _ = path.split(var_name + \"=\")[1]\n",
    "    return _.split(\"/\")[0]\n",
    "\n",
    "\n",
    "def load_gene_stats_malignant_cibersortx():\n",
    "    df = pd.concat(\n",
    "        {str(path): pd.read_parquet(path) for path in parquet_paths},\n",
    "        names=[\"path\", \"index\"],\n",
    "    )\n",
    "    df[\"run_id\"] = df.index.get_level_values(\"path\").map(\n",
    "        lambda path: extract_from_path(path, \"run_id\")\n",
    "    )\n",
    "    df[\"malignant_means\"] = df.index.get_level_values(\"path\").map(\n",
    "        lambda path: extract_from_path(path, \"malignant_means\")\n",
    "    )\n",
    "    # df = df.droplevel([\"path\", \"index\"])\n",
    "    df = df.set_index([\"run_id\", \"malignant_means\", \"gene_symbol\"])\n",
    "    return df\n",
    "\n",
    "\n",
    "gene_stats_malignant_cibersortx = load_gene_stats_malignant_cibersortx()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>pval</th>\n",
       "      <th>fold_change</th>\n",
       "      <th>sparsity_overall</th>\n",
       "      <th>-log10_pval</th>\n",
       "      <th>log2_fold_change</th>\n",
       "      <th>-log10_pval_signed</th>\n",
       "      <th>significant_bh_fdr=0.10</th>\n",
       "      <th>significant_bh_fdr=0.25</th>\n",
       "      <th>perturbed</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>run_id</th>\n",
       "      <th>malignant_means</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">00</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">0.55,0.85</th>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.045960</td>\n",
       "      <td>0.762946</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.337620</td>\n",
       "      <td>-0.390348</td>\n",
       "      <td>-1.337620</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.671588</td>\n",
       "      <td>0.979047</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.172897</td>\n",
       "      <td>-0.030550</td>\n",
       "      <td>-0.172897</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.006673</td>\n",
       "      <td>0.734658</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.175654</td>\n",
       "      <td>-0.444854</td>\n",
       "      <td>-2.175654</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">09</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">0.71,0.71</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>0.142938</td>\n",
       "      <td>1.115149</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.844852</td>\n",
       "      <td>0.157237</td>\n",
       "      <td>0.844852</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.121704</td>\n",
       "      <td>1.105519</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.914697</td>\n",
       "      <td>0.144724</td>\n",
       "      <td>0.914697</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0.148667</td>\n",
       "      <td>1.112070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.827785</td>\n",
       "      <td>0.153247</td>\n",
       "      <td>0.827785</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.336206</td>\n",
       "      <td>1.119979</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.473394</td>\n",
       "      <td>0.163472</td>\n",
       "      <td>0.473394</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.574221</td>\n",
       "      <td>0.927400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.240921</td>\n",
       "      <td>-0.108736</td>\n",
       "      <td>-0.240921</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>803150 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        pval  fold_change  sparsity_overall  \\\n",
       "run_id malignant_means gene_symbol                                            \n",
       "00     0.55,0.85       A1BG         0.045960     0.762946               0.0   \n",
       "                       A2M          0.671588     0.979047               0.0   \n",
       "                       A2ML1        0.006673     0.734658               0.0   \n",
       "                       A4GALT       1.000000     1.000000               0.0   \n",
       "                       A4GNT        1.000000     1.000000               0.0   \n",
       "...                                      ...          ...               ...   \n",
       "09     0.71,0.71       ZYG11A       0.142938     1.115149               0.0   \n",
       "                       ZYG11B       0.121704     1.105519               0.0   \n",
       "                       ZYX          0.148667     1.112070               0.0   \n",
       "                       ZZEF1        0.336206     1.119979               0.0   \n",
       "                       ZZZ3         0.574221     0.927400               0.0   \n",
       "\n",
       "                                    -log10_pval  log2_fold_change  \\\n",
       "run_id malignant_means gene_symbol                                  \n",
       "00     0.55,0.85       A1BG            1.337620         -0.390348   \n",
       "                       A2M             0.172897         -0.030550   \n",
       "                       A2ML1           2.175654         -0.444854   \n",
       "                       A4GALT         -0.000000          0.000000   \n",
       "                       A4GNT          -0.000000          0.000000   \n",
       "...                                         ...               ...   \n",
       "09     0.71,0.71       ZYG11A          0.844852          0.157237   \n",
       "                       ZYG11B          0.914697          0.144724   \n",
       "                       ZYX             0.827785          0.153247   \n",
       "                       ZZEF1           0.473394          0.163472   \n",
       "                       ZZZ3            0.240921         -0.108736   \n",
       "\n",
       "                                    -log10_pval_signed  \\\n",
       "run_id malignant_means gene_symbol                       \n",
       "00     0.55,0.85       A1BG                  -1.337620   \n",
       "                       A2M                   -0.172897   \n",
       "                       A2ML1                 -2.175654   \n",
       "                       A4GALT                -0.000000   \n",
       "                       A4GNT                 -0.000000   \n",
       "...                                                ...   \n",
       "09     0.71,0.71       ZYG11A                 0.844852   \n",
       "                       ZYG11B                 0.914697   \n",
       "                       ZYX                    0.827785   \n",
       "                       ZZEF1                  0.473394   \n",
       "                       ZZZ3                  -0.240921   \n",
       "\n",
       "                                    significant_bh_fdr=0.10  \\\n",
       "run_id malignant_means gene_symbol                            \n",
       "00     0.55,0.85       A1BG                           False   \n",
       "                       A2M                            False   \n",
       "                       A2ML1                           True   \n",
       "                       A4GALT                         False   \n",
       "                       A4GNT                          False   \n",
       "...                                                     ...   \n",
       "09     0.71,0.71       ZYG11A                         False   \n",
       "                       ZYG11B                         False   \n",
       "                       ZYX                            False   \n",
       "                       ZZEF1                          False   \n",
       "                       ZZZ3                           False   \n",
       "\n",
       "                                    significant_bh_fdr=0.25  perturbed  \n",
       "run_id malignant_means gene_symbol                                      \n",
       "00     0.55,0.85       A1BG                            True      False  \n",
       "                       A2M                            False      False  \n",
       "                       A2ML1                           True      False  \n",
       "                       A4GALT                         False      False  \n",
       "                       A4GNT                          False      False  \n",
       "...                                                     ...        ...  \n",
       "09     0.71,0.71       ZYG11A                         False      False  \n",
       "                       ZYG11B                         False      False  \n",
       "                       ZYX                            False      False  \n",
       "                       ZZEF1                          False      False  \n",
       "                       ZZZ3                           False      False  \n",
       "\n",
       "[803150 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_stats_malignant_cibersortx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>pval</th>\n",
       "      <th>sparsity_overall</th>\n",
       "      <th>-log10_pval</th>\n",
       "      <th>log2_fold_change</th>\n",
       "      <th>significant_bh_fdr=0.10</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>malignant_means</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">0.55,0.85</th>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.559052</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.356863</td>\n",
       "      <td>0.059536</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.410739</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.482232</td>\n",
       "      <td>-0.144573</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.132443</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.466375</td>\n",
       "      <td>-0.363758</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GNT</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">0.71,0.71</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <td>0.685009</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.224542</td>\n",
       "      <td>0.049764</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYG11B</th>\n",
       "      <td>0.633558</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.317697</td>\n",
       "      <td>-0.016852</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZYX</th>\n",
       "      <td>0.543524</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.433230</td>\n",
       "      <td>0.002201</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZEF1</th>\n",
       "      <td>0.456068</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.619656</td>\n",
       "      <td>0.096869</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZZZ3</th>\n",
       "      <td>0.662656</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.195178</td>\n",
       "      <td>-0.018776</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>80315 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 pval  sparsity_overall  -log10_pval  \\\n",
       "malignant_means gene_symbol                                            \n",
       "0.55,0.85       A1BG         0.559052               0.0     0.356863   \n",
       "                A2M          0.410739               0.0     0.482232   \n",
       "                A2ML1        0.132443               0.0     1.466375   \n",
       "                A4GALT       1.000000               0.0     0.000000   \n",
       "                A4GNT        1.000000               0.0     0.000000   \n",
       "...                               ...               ...          ...   \n",
       "0.71,0.71       ZYG11A       0.685009               0.0     0.224542   \n",
       "                ZYG11B       0.633558               0.0     0.317697   \n",
       "                ZYX          0.543524               0.0     0.433230   \n",
       "                ZZEF1        0.456068               0.0     0.619656   \n",
       "                ZZZ3         0.662656               0.0     0.195178   \n",
       "\n",
       "                             log2_fold_change  significant_bh_fdr=0.10  \n",
       "malignant_means gene_symbol                                             \n",
       "0.55,0.85       A1BG                 0.059536                      0.0  \n",
       "                A2M                 -0.144573                      0.0  \n",
       "                A2ML1               -0.363758                      0.3  \n",
       "                A4GALT               0.000000                      0.0  \n",
       "                A4GNT                0.000000                      0.0  \n",
       "...                                       ...                      ...  \n",
       "0.71,0.71       ZYG11A               0.049764                      0.0  \n",
       "                ZYG11B              -0.016852                      0.0  \n",
       "                ZYX                  0.002201                      0.0  \n",
       "                ZZEF1                0.096869                      0.0  \n",
       "                ZZZ3                -0.018776                      0.0  \n",
       "\n",
       "[80315 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_gene_stats_malignant_cibersortx_aggregated(df):\n",
    "    df = df.groupby([\"malignant_means\", \"gene_symbol\"])[\n",
    "        [\"pval\", \"sparsity_overall\", \"-log10_pval\", \"log2_fold_change\", \"significant_bh_fdr=0.10\"]\n",
    "    ].agg(\"mean\")\n",
    "    return df\n",
    "\n",
    "\n",
    "gene_stats_malignant_cibersortx_aggregated = load_gene_stats_malignant_cibersortx_aggregated(\n",
    "    gene_stats_malignant_cibersortx\n",
    ")\n",
    "\n",
    "gene_stats_malignant_cibersortx_aggregated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "Unable to allocate 1.27 GiB for an array with shape (7186, 23686) and data type float64",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m df_sc_rnaseq, df_sc_metadata \u001b[39m=\u001b[39m helpers\u001b[39m.\u001b[39;49mdatasets\u001b[39m.\u001b[39;49mload_jerby_arnon_hg19_tpm()\n",
      "File \u001b[0;32m~/deconv/helpers/datasets/jerby_arnon.py:38\u001b[0m, in \u001b[0;36mload_jerby_arnon_hg19_tpm\u001b[0;34m()\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Load Jerby-Arnon scRNA-seq data (hg19 tpm) from GEO (GSE115978)\u001b[39;00m\n\u001b[1;32m     26\u001b[0m \n\u001b[1;32m     27\u001b[0m \u001b[39m:return: sc_hg19_tpm, metadata\u001b[39;00m\n\u001b[1;32m     28\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m     29\u001b[0m logger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mloading Jerby-Arnon scRNA-seq data\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     30\u001b[0m sc_hg19_tpm \u001b[39m=\u001b[39m (\n\u001b[1;32m     31\u001b[0m     pd\u001b[39m.\u001b[39;49mread_csv(\n\u001b[1;32m     32\u001b[0m         \u001b[39m\"\u001b[39;49m\u001b[39mgs://liulab/ftp/GSE115978/GSE115978_tpm.csv\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m     33\u001b[0m         index_col\u001b[39m=\u001b[39;49m\u001b[39m0\u001b[39;49m,\n\u001b[1;32m     34\u001b[0m         engine\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mpyarrow\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m     35\u001b[0m         \u001b[39m# skiprows=lambda i: i % 25,\u001b[39;49;00m\n\u001b[1;32m     36\u001b[0m     )\n\u001b[1;32m     37\u001b[0m     \u001b[39m.\u001b[39;49mrename_axis(index\u001b[39m=\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mGENE_SYMBOL, columns\u001b[39m=\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mSINGLE_CELL_ID)\n\u001b[0;32m---> 38\u001b[0m     \u001b[39m.\u001b[39;49msort_index(axis\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mcolumns\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m     39\u001b[0m     \u001b[39m.\u001b[39msort_index(axis\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mrows\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     40\u001b[0m )\n\u001b[1;32m     41\u001b[0m logger\u001b[39m.\u001b[39mdebug(\u001b[39m\"\u001b[39m\u001b[39mloading Jerby-Arnon metadata\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m     42\u001b[0m metadata \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mread_csv(\n\u001b[1;32m     43\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mgs://liulab/ftp/GSE115978/GSE115978_cell.annotations.csv\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     44\u001b[0m     na_values\u001b[39m=\u001b[39m{\u001b[39m\"\u001b[39m\u001b[39mcell.types\u001b[39m\u001b[39m\"\u001b[39m: \u001b[39m\"\u001b[39m\u001b[39m?\u001b[39m\u001b[39m\"\u001b[39m},\n\u001b[1;32m     45\u001b[0m )\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/util/_decorators.py:331\u001b[0m, in \u001b[0;36mdeprecate_nonkeyword_arguments.<locals>.decorate.<locals>.wrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    325\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(args) \u001b[39m>\u001b[39m num_allow_args:\n\u001b[1;32m    326\u001b[0m     warnings\u001b[39m.\u001b[39mwarn(\n\u001b[1;32m    327\u001b[0m         msg\u001b[39m.\u001b[39mformat(arguments\u001b[39m=\u001b[39m_format_argument_list(allow_args)),\n\u001b[1;32m    328\u001b[0m         \u001b[39mFutureWarning\u001b[39;00m,\n\u001b[1;32m    329\u001b[0m         stacklevel\u001b[39m=\u001b[39mfind_stack_level(),\n\u001b[1;32m    330\u001b[0m     )\n\u001b[0;32m--> 331\u001b[0m \u001b[39mreturn\u001b[39;00m func(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/frame.py:7095\u001b[0m, in \u001b[0;36mDataFrame.sort_index\u001b[0;34m(self, axis, level, ascending, inplace, kind, na_position, sort_remaining, ignore_index, key)\u001b[0m\n\u001b[1;32m   6993\u001b[0m \u001b[39m@deprecate_nonkeyword_arguments\u001b[39m(version\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, allowed_args\u001b[39m=\u001b[39m[\u001b[39m\"\u001b[39m\u001b[39mself\u001b[39m\u001b[39m\"\u001b[39m])\n\u001b[1;32m   6994\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39msort_index\u001b[39m(  \u001b[39m# type: ignore[override]\u001b[39;00m\n\u001b[1;32m   6995\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   7004\u001b[0m     key: IndexKeyFunc \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m,\n\u001b[1;32m   7005\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m DataFrame \u001b[39m|\u001b[39m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m   7006\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m   7007\u001b[0m \u001b[39m    Sort object by labels (along an axis).\u001b[39;00m\n\u001b[1;32m   7008\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   7093\u001b[0m \u001b[39m    d  4\u001b[39;00m\n\u001b[1;32m   7094\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 7095\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49msort_index(\n\u001b[1;32m   7096\u001b[0m         axis\u001b[39m=\u001b[39;49maxis,\n\u001b[1;32m   7097\u001b[0m         level\u001b[39m=\u001b[39;49mlevel,\n\u001b[1;32m   7098\u001b[0m         ascending\u001b[39m=\u001b[39;49mascending,\n\u001b[1;32m   7099\u001b[0m         inplace\u001b[39m=\u001b[39;49minplace,\n\u001b[1;32m   7100\u001b[0m         kind\u001b[39m=\u001b[39;49mkind,\n\u001b[1;32m   7101\u001b[0m         na_position\u001b[39m=\u001b[39;49mna_position,\n\u001b[1;32m   7102\u001b[0m         sort_remaining\u001b[39m=\u001b[39;49msort_remaining,\n\u001b[1;32m   7103\u001b[0m         ignore_index\u001b[39m=\u001b[39;49mignore_index,\n\u001b[1;32m   7104\u001b[0m         key\u001b[39m=\u001b[39;49mkey,\n\u001b[1;32m   7105\u001b[0m     )\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/generic.py:5028\u001b[0m, in \u001b[0;36mNDFrame.sort_index\u001b[0;34m(self, axis, level, ascending, inplace, kind, na_position, sort_remaining, ignore_index, key)\u001b[0m\n\u001b[1;32m   5025\u001b[0m         \u001b[39mreturn\u001b[39;00m result\n\u001b[1;32m   5027\u001b[0m baxis \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_get_block_manager_axis(axis)\n\u001b[0;32m-> 5028\u001b[0m new_data \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_mgr\u001b[39m.\u001b[39;49mtake(indexer, axis\u001b[39m=\u001b[39;49mbaxis, verify\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m)\n\u001b[1;32m   5030\u001b[0m \u001b[39m# reconstruct axis if needed\u001b[39;00m\n\u001b[1;32m   5031\u001b[0m new_data\u001b[39m.\u001b[39mset_axis(baxis, new_data\u001b[39m.\u001b[39maxes[baxis]\u001b[39m.\u001b[39m_sort_levels_monotonic())\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/internals/managers.py:978\u001b[0m, in \u001b[0;36mBaseBlockManager.take\u001b[0;34m(self, indexer, axis, verify, convert_indices)\u001b[0m\n\u001b[1;32m    975\u001b[0m     indexer \u001b[39m=\u001b[39m maybe_convert_indices(indexer, n, verify\u001b[39m=\u001b[39mverify)\n\u001b[1;32m    977\u001b[0m new_labels \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39maxes[axis]\u001b[39m.\u001b[39mtake(indexer)\n\u001b[0;32m--> 978\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mreindex_indexer(\n\u001b[1;32m    979\u001b[0m     new_axis\u001b[39m=\u001b[39;49mnew_labels,\n\u001b[1;32m    980\u001b[0m     indexer\u001b[39m=\u001b[39;49mindexer,\n\u001b[1;32m    981\u001b[0m     axis\u001b[39m=\u001b[39;49maxis,\n\u001b[1;32m    982\u001b[0m     allow_dups\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m,\n\u001b[1;32m    983\u001b[0m     copy\u001b[39m=\u001b[39;49m\u001b[39mNone\u001b[39;49;00m,\n\u001b[1;32m    984\u001b[0m )\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/internals/managers.py:743\u001b[0m, in \u001b[0;36mBaseBlockManager.reindex_indexer\u001b[0;34m(self, new_axis, indexer, axis, fill_value, allow_dups, copy, only_slice, use_na_proxy)\u001b[0m\n\u001b[1;32m    740\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mIndexError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mRequested axis not found in manager\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    742\u001b[0m \u001b[39mif\u001b[39;00m axis \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[0;32m--> 743\u001b[0m     new_blocks, new_refs \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_slice_take_blocks_ax0(\n\u001b[1;32m    744\u001b[0m         indexer,\n\u001b[1;32m    745\u001b[0m         fill_value\u001b[39m=\u001b[39;49mfill_value,\n\u001b[1;32m    746\u001b[0m         only_slice\u001b[39m=\u001b[39;49monly_slice,\n\u001b[1;32m    747\u001b[0m         use_na_proxy\u001b[39m=\u001b[39;49muse_na_proxy,\n\u001b[1;32m    748\u001b[0m     )\n\u001b[1;32m    749\u001b[0m     parent \u001b[39m=\u001b[39m \u001b[39mNone\u001b[39;00m \u001b[39mif\u001b[39;00m com\u001b[39m.\u001b[39mall_none(\u001b[39m*\u001b[39mnew_refs) \u001b[39melse\u001b[39;00m \u001b[39mself\u001b[39m\n\u001b[1;32m    750\u001b[0m \u001b[39melse\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/internals/managers.py:912\u001b[0m, in \u001b[0;36mBaseBlockManager._slice_take_blocks_ax0\u001b[0;34m(self, slice_or_indexer, fill_value, only_slice, use_na_proxy)\u001b[0m\n\u001b[1;32m    910\u001b[0m         refs\u001b[39m.\u001b[39mappend(weakref\u001b[39m.\u001b[39mref(blk))\n\u001b[1;32m    911\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 912\u001b[0m     nb \u001b[39m=\u001b[39m blk\u001b[39m.\u001b[39;49mtake_nd(taker, axis\u001b[39m=\u001b[39;49m\u001b[39m0\u001b[39;49m, new_mgr_locs\u001b[39m=\u001b[39;49mmgr_locs)\n\u001b[1;32m    913\u001b[0m     blocks\u001b[39m.\u001b[39mappend(nb)\n\u001b[1;32m    914\u001b[0m     refs\u001b[39m.\u001b[39mappend(\u001b[39mNone\u001b[39;00m)\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/internals/blocks.py:880\u001b[0m, in \u001b[0;36mBlock.take_nd\u001b[0;34m(self, indexer, axis, new_mgr_locs, fill_value)\u001b[0m\n\u001b[1;32m    877\u001b[0m     allow_fill \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m    879\u001b[0m \u001b[39m# Note: algos.take_nd has upcast logic similar to coerce_to_target_dtype\u001b[39;00m\n\u001b[0;32m--> 880\u001b[0m new_values \u001b[39m=\u001b[39m algos\u001b[39m.\u001b[39;49mtake_nd(\n\u001b[1;32m    881\u001b[0m     values, indexer, axis\u001b[39m=\u001b[39;49maxis, allow_fill\u001b[39m=\u001b[39;49mallow_fill, fill_value\u001b[39m=\u001b[39;49mfill_value\n\u001b[1;32m    882\u001b[0m )\n\u001b[1;32m    884\u001b[0m \u001b[39m# Called from three places in managers, all of which satisfy\u001b[39;00m\n\u001b[1;32m    885\u001b[0m \u001b[39m#  this assertion\u001b[39;00m\n\u001b[1;32m    886\u001b[0m \u001b[39massert\u001b[39;00m \u001b[39mnot\u001b[39;00m (axis \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m \u001b[39mand\u001b[39;00m new_mgr_locs \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m)\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/array_algos/take.py:117\u001b[0m, in \u001b[0;36mtake_nd\u001b[0;34m(arr, indexer, axis, fill_value, allow_fill)\u001b[0m\n\u001b[1;32m    114\u001b[0m     \u001b[39mreturn\u001b[39;00m arr\u001b[39m.\u001b[39mtake(indexer, fill_value\u001b[39m=\u001b[39mfill_value, allow_fill\u001b[39m=\u001b[39mallow_fill)\n\u001b[1;32m    116\u001b[0m arr \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39masarray(arr)\n\u001b[0;32m--> 117\u001b[0m \u001b[39mreturn\u001b[39;00m _take_nd_ndarray(arr, indexer, axis, fill_value, allow_fill)\n",
      "File \u001b[0;32m/opt/conda/envs/deconv/lib/python3.11/site-packages/pandas/core/array_algos/take.py:158\u001b[0m, in \u001b[0;36m_take_nd_ndarray\u001b[0;34m(arr, indexer, axis, fill_value, allow_fill)\u001b[0m\n\u001b[1;32m    156\u001b[0m     out \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mempty(out_shape, dtype\u001b[39m=\u001b[39mdtype, order\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mF\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    157\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 158\u001b[0m     out \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mempty(out_shape, dtype\u001b[39m=\u001b[39mdtype)\n\u001b[1;32m    160\u001b[0m func \u001b[39m=\u001b[39m _get_take_nd_function(\n\u001b[1;32m    161\u001b[0m     arr\u001b[39m.\u001b[39mndim, arr\u001b[39m.\u001b[39mdtype, out\u001b[39m.\u001b[39mdtype, axis\u001b[39m=\u001b[39maxis, mask_info\u001b[39m=\u001b[39mmask_info\n\u001b[1;32m    162\u001b[0m )\n\u001b[1;32m    163\u001b[0m func(arr, indexer, out, fill_value)\n",
      "\u001b[0;31mMemoryError\u001b[0m: Unable to allocate 1.27 GiB for an array with shape (7186, 23686) and data type float64"
     ]
    }
   ],
   "source": [
    "df_sc_rnaseq, df_sc_metadata = helpers.datasets.load_jerby_arnon_hg19_tpm()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_gene_expression_ratios(df_sc_rnaseq: pd.DataFrame, df_sc_metadata: pd.DataFrame):\n",
    "    df = df_sc_metadata[[\"cell_type\"]]\n",
    "    # drop cells with missing cell_type\n",
    "    df = df.dropna()\n",
    "    # drop cells with cell_type \"CAF\" or \"Endothelial\"\n",
    "    df = df[~df[\"cell_type\"].isin([\"CAF\", \"Endothelial\"])]\n",
    "    # add a string column with value \"malignant\" if cell type is \"Malignant\" or \"other\" otherwise\n",
    "    df[\"cell_group\"] = df[\"cell_type\"].map(lambda x: \"malignant\" if x == \"Malignant\" else \"other\")\n",
    "    # print(df)\n",
    "\n",
    "    # add gene expression for each single_cell_id\n",
    "    df_sc_rnaseq = pd.merge(df[\"cell_group\"], df_sc_rnaseq.T, left_index=True, right_index=True)\n",
    "    df = df_sc_rnaseq.groupby(\"cell_group\").mean()\n",
    "    df = df.rename_axis(columns=\"gene_symbol\").T\n",
    "    df[\"malignant_ratio\"] = df[\"malignant\"] / df[\"other\"]\n",
    "    df = df.sort_values(\"malignant_ratio\", ascending=False)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_gene_data(df_sc_rnaseq: pd.DataFrame, gene_stats: pd.DataFrame) -> pd.DataFrame:\n",
    "    # genes = gene_stats.index.get_level_values(\"gene_symbol\").unique()\n",
    "    # filter out genes not in df_sc_rnaseq\n",
    "    # genes = genes[genes.isin(df_sc_rnaseq.index)]\n",
    "    # df_sc_rnaseq = df_sc_rnaseq.loc[genes]\n",
    "    df_sc_rnaseq = df_sc_rnaseq.loc[\n",
    "        df_sc_rnaseq.index.isin(gene_stats.index.get_level_values(\"gene_symbol\"))\n",
    "    ]\n",
    "    gene_expression_ratios = compute_gene_expression_ratios(df_sc_rnaseq, df_sc_metadata)\n",
    "    return pd.merge(gene_expression_ratios, gene_stats, on=\"gene_symbol\")\n",
    "\n",
    "\n",
    "df_gene_data = compute_all(df_sc_rnaseq)\n",
    "\n",
    "df_gene_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# number of genes per malignant_means\n",
    "df_gene_data.groupby(\"malignant_means\").size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.express as px"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_scatter(\n",
    "    df_gene_data: pd.DataFrame,\n",
    "):\n",
    "    df_gene_data = df_gene_data.sort_index(ascending=[False, True])\n",
    "    df_gene_data = df_gene_data.reset_index()\n",
    "    fig = px.scatter(\n",
    "        df_gene_data,\n",
    "        x=\"malignant_ratio\",\n",
    "        # y=\"significant_bh_fdr=0.10\",\n",
    "        y=\"-log10_pval\",\n",
    "        hover_name=\"gene_symbol\",\n",
    "        hover_data=df_gene_data.columns,\n",
    "        color=\"\"\n",
    "        facet_col=\"malignant_means\",\n",
    "    )\n",
    "    fig.update_xaxes(type=\"log\")\n",
    "    fig.update_layout(\n",
    "        title=\"Gene expression ratio vs. CIBERSORTx significance\",\n",
    "        xaxis_title=\"Malignant / Other\",\n",
    "        yaxis_title=\"CIBERSORTx significance\",\n",
    "    )\n",
    "    # add histograms along y-axis\n",
    "    return fig\n",
    "\n",
    "fig = make_scatter(df_gene_data)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'pd' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m# density contour of pval and ratio\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mmake_plot\u001b[39m(\n\u001b[0;32m----> 3\u001b[0m     df_gene_data: pd\u001b[39m.\u001b[39mDataFrame\n\u001b[1;32m      4\u001b[0m ):\n\u001b[1;32m      5\u001b[0m     df_gene_data \u001b[39m=\u001b[39m df_gene_data\u001b[39m.\u001b[39msort_index(ascending\u001b[39m=\u001b[39m[\u001b[39mFalse\u001b[39;00m, \u001b[39mTrue\u001b[39;00m])\n\u001b[1;32m      6\u001b[0m     df \u001b[39m=\u001b[39m df_gene_data\u001b[39m.\u001b[39mreset_index()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'pd' is not defined"
     ]
    }
   ],
   "source": [
    "# density contour of pval and ratio\n",
    "def make_plot(df_gene_data: pd.DataFrame):\n",
    "    df_gene_data = df_gene_data.sort_index(ascending=[False, True])\n",
    "    df = df_gene_data.reset_index()\n",
    "    fig = px.density_contour(\n",
    "        df,\n",
    "        x=\"malignant_ratio\",\n",
    "        y=\"-log10_pval\",\n",
    "    )\n",
    "    return fig\n",
    "\n",
    "\n",
    "fig = make_plot(df_gene_data)\n",
    "fig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "kernel": "python3",
   "name": "jax-gpu.0-3-14.m96",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/jax-gpu.0-3-14:m96"
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
